src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,327 @@
+                    {
+                        using var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, _runningQuery.Token);
+                        {
+                            await using (var stream =
+                                await Dht.SwarmService.DialAsync(peer, Dht.ToString(), cts.Token).ConfigureAwait(false))
+                            {
+                                // Send the KAD query and get a response.
+                                Serializer.SerializeWithLengthPrefix(stream, _queryMessage, PrefixStyle.Base128);
+                                await stream.FlushAsync(cts.Token).ConfigureAwait(false);
+                                var response = await ProtoBufHelper.ReadMessageAsync<DhtMessage>(stream, cts.Token)
+                                   .ConfigureAwait(false);
+
+                                // Process answer
+                                ProcessProviders(response.ProviderPeers);
+                                ProcessCloserPeers(response.CloserPeers);
+                            }
+                        }
+                    }
+
+                    var time = DateTime.Now - start;
+                    Log.Debug($"Q{Id}.{taskId}.{pass} ok {peer} ({time.TotalMilliseconds} ms)");
+                }
+                catch (Exception e)
+                {
+                    Interlocked.Increment(ref _failedConnects);
+                    var time = DateTime.Now - start;
+                    Log.Warn($"Q{Id}.{taskId}.{pass} failed ({time.TotalMilliseconds} ms) - {e.Message}");
+                }
+                finally
+                {
+                    AskCount.Release();
+                }
+            }
+        }
+
+        private void ProcessProviders(DhtPeerMessage[] providers)
+        {
+            if (providers == null)
+            {
+                return;
+            }
+
+            foreach (var provider in providers)
+                if (provider.TryToPeer(out var p))
+                {
+                    if (p == Dht.SwarmService.LocalPeer || !Dht.SwarmService.IsAllowed(p))
+                    {
+                        continue;
+                    }
+
+                    p = Dht.SwarmService.RegisterPeer(p);
+                    if (QueryType != MessageType.GetProviders)
+                    {
+                        continue;
+                    }
+
+                    // Only unique answers
+                    var answer = p as T;
+                    if (!_answers.ContainsKey(answer ?? throw new NullReferenceException()))
+                    {
+                        AddAnswer(answer);
+                    }
+                }
+        }
+
+        private void ProcessCloserPeers(DhtPeerMessage[] closerPeers)
+        {
+            if (closerPeers == null)
+            {
+                return;
+            }
+            
+            foreach (var closer in closerPeers)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)